### ISSUE TITLE
Clarifications Billing Plans and Agreement parameters

### ISSUE THREAD

**Issue by knifesk** [paypal/PayPal-REST-API-issues#68]
> For internal bookkeeping: DEVDOC-666

### General information

* Environment: Sandbox
* Language, language version, and OS: PHP 7 on Ubuntu 16.04

### Issue description

Sandbox payment events for agreements are not processing immediately. In fact, it can take several hours for PAYMENT.SALE.COMPLETED webhook event to be sent and paypal funds are not discounted until that happens. Is this normal? I created several plans and agreements and I receive the webhook "BILLING.SUBSCRIPTION.CREATED" right away.. but payments takes way too long... this behaviour started happening a couple weeks ago... before that I was getting the webhook right after BILLING.SUBSCRIPTION.CREATED

**Comment by xiaoleih41** [comment]
I am looking into this issue. If you can provide a debug ID, it will be much more helpful.

**Comment by knifesk** [comment]
I just created a new subscription, received the created hook, but not the payment. Here's the body of the hook I received.

```json
{
    "id": "WH-9F8512228R455853D-76350277XA663505S",
    "event_version": "1.0",
    "create_time": "2017-04-26T15:26:48.112Z",
    "resource_type": "Agreement",
    "event_type": "BILLING.SUBSCRIPTION.CREATED",
    "summary": "A billing subscription was created",
    "resource": {
        "agreement_details": {
            "outstanding_balance": {
                "value": "0.00"
            },
            "num_cycles_remaining": "0",
            "num_cycles_completed": "0",
            "next_billing_date": "2017-04-26T10:00:00Z",
            "final_payment_due_date": "1970-01-01T00:00:00Z",
            "failed_payment_count": "0"
        },
        "description": "asdasdasdasdasdas",
        "links": [
            {
                "href": "api.sandbox.paypal.com/v1/payments/billing-agreements/I-2158E8N8C1LW",
                "rel": "self",
                "method": "GET"
            }
        ],
        "shipping_address": {
            "recipient_name": "Jon Snow",
            "line1": "001",
            "city": "San Jose",
            "state": "CA",
            "postal_code": "50002",
            "country_code": "US"
        },
        "id": "I-2158E8N8C1LW",
        "state": "Active",
        "payer": {
            "payment_method": "paypal",
            "status": "verified",
            "payer_info": {
                "email": "jsnow@gmail.com",
                "first_name": "Jon",
                "last_name": "Snow",
                "payer_id": "XAQQM537J65RN",
                "shipping_address": {
                    "recipient_name": "Jon Snow",
                    "line1": "001",
                    "city": "San Jose",
                    "state": "CA",
                    "postal_code": "50002",
                    "country_code": "US"
                }
            }
        },
        "plan": {
            "curr_code": "USD",
            "links": [],
            "payment_definitions": [
                {
                    "type": "REGULAR",
                    "frequency": "Month",
                    "frequency_interval": "1",
                    "amount": {
                        "value": "30.00"
                    },
                    "cycles": "0",
                    "charge_models": [
                        {
                            "type": "TAX",
                            "amount": {
                                "value": "0.00"
                            }
                        },
                        {
                            "type": "SHIPPING",
                            "amount": {
                                "value": "0.00"
                            }
                        }
                    ]
                }
            ],
            "merchant_preferences": {
                "setup_fee": {
                    "value": "0.00"
                },
                "auto_bill_amount": "YES",
                "max_fail_attempts": "2"
            }
        },
        "start_date": "2017-04-26T07:00:00Z"
    },
    "links": [
        {
            "href": "https://api.sandbox.paypal.com/v1/notifications/webhooks-events/WH-9F8512228R455853D-76350277XA663505S",
            "rel": "self",
            "method": "GET"
        },
        {
            "href": "https://api.sandbox.paypal.com/v1/notifications/webhooks-events/WH-9F8512228R455853D-76350277XA663505S/resend",
            "rel": "resend",
            "method": "POST"
        }
    ]
}
```

**Comment by xiaoleih41** [comment]
You didn't set the setup fee which will charge the buyer instantly while the billing agreement is approved. Therefore the first recurring payment will be debited based on the recurring payment batch that it will be queued in.  Depending on the batch, delay could be expected.

**Comment by pp-randy** [comment]
To add to what @xiaoleih41 has said in regards to setup_fee. 

We need to follow up on the documentation to help clear things up, because `start_date` is going to be truncated to the beginning of the day for your accounts timezone, but is displayed as as UTC, which is why it shows as 7:00 AM but is also midnight in PDT.  This is no longer a future date.

**Comment by knifesk** [comment]
Thanks for the quick replies guys! much appreciated!

While it's true that I didn't configure setup fees in this subscription, I'm almost sure that a month ago I was getting the subscription created, the setup fee and the payment hooks immediately... 

Is it possible that the subscription is created and then the first payment is rejected? I'm waiting on the payment to confirm the user account because of a possible rejection.. but if that's not possible I can confirm the user account in the billing created event. 

My second problem is that while the payment hasn't arrived, the user could switch to a another plan. This is  highly unlikely, but could happen. The process of plan switching is the following: refund the last payment (partially), cancel the subscription and then create a new one.. If no payment has arrived I can not make the refund, but if the notification arrives later anyway that could cause a problem. Unless the hook is de-queued if the subscription is cancelled.

Regarding to `start_date`. I'm setting the start date 2 minutes in the future of "now"... 
```php
        $now = new Datetime();
        $now->add(new \DateInterval('PT2M'));

        $agreement = new Agreement();

        $agreement
                ->setName($nombre)
                ->setDescription($descripcion)
                ->setStartDate($now->format(Datetime::ISO8601));
```
Subscription are immediately, as soon as you confirm the subscription and the created hook arrives the account data are sent via email, but disabled until payment arrives.. and sometimes, the payment takes hours to do so. This leaves my customers blocked out of their account until the payment is cleared.

Advices?

**Comment by pp-randy** [comment]
Start_date is truncated to midnight, so adding 2 minutes won't do much.   I don't think using today's date over setup_fee is the way to go honestly.  

> My second problem is that while the payment hasn't arrived, the user could switch to a another plan.
> This leaves my customers blocked out of their account until the payment is cleared.

Wouldn't using setup_fee also help with both of  these problems? 

We cleaned up a bunch of sandbox recurring payments a few months back that were causing delays.

**Comment by knifesk** [comment]
So, you say that I should charge the first payment on setup_fee and set the
start date after the first billing cycle? For example: subscription is
30USD/month. Set start date in 30 days and charge the first 30USD with
setup fee?

**Comment by pp-randy** [comment]
To clarify setup_fee is the initial payment if its used, it will show in the history.  Give it the amount of the first billing.

**Comment by pp-randy** [comment]
Sorry I didn't see your questions I was just adding another comment, 

I think I answered your question with it.  Yes, to your questions. 

The setup_fee since can be used with the [`initial_fail_amount_action`](https://developer.paypal.com/docs/api/payments.billing-plans#definition-merchant_preferences) setting as well.

**Comment by knifesk** [comment]
I interpreted that `setup_fee` was an one-time additional charge apart from the first billing. Maybe I'm getting all wrong.

Concluding: 
- Delay are normal in the first payment. It is not immediate.
- If I need immediate payment I should set the `setup_fee` to the value of the first payment and set `start_date` to the date corresponding to the second payment.
- If I needed extra one-time charge for setup, I should set `setup_fee` to the cost of the subscription + this additional charge

Now a new question comes to my head? If I set the `start_date` 30 days in the future.. Do I get the subscription.created hook anyway?

**Comment by pp-randy** [comment]
So in our Classic API `setup_fee` was called `INITAMT` https://developer.paypal.com/docs/classic/api/merchant/CreateRecurringPaymentsProfile_API_Operation_NVP/
![image](https://cloud.githubusercontent.com/assets/6819909/25453673/7d8a727e-2a8f-11e7-8711-ee2adfd4aca0.png)

Also other things seem to be documented here that we should get on the REST API documentation.
![image](https://cloud.githubusercontent.com/assets/6819909/25453617/4034028c-2a8f-11e7-94e5-d86c47705bb5.png)

Yes, you still get the subscription.created webhook at the time its created independent of `start_date`.

**Comment by pp-randy** [comment]
I have updated the initial comment with internal tracking information regarding documentation updates.

**Comment by pp-randy** [comment]
The documentation has been updated.  
https://developer.paypal.com/docs/api/payments.billing-agreements#agreement_create_request
https://developer.paypal.com/docs/api/payments.billing-plans#definition-merchant_preferences
